# Improvisation of Accuracy on Zomato India Data Set

In [52]:
import pandas as pd
df = pd.read_csv('ZomatoIndia.csv',encoding = 'latin-1', sep = ',') # Reading the csv file
df.head()

Unnamed: 0  Restaurant ID                        Restaurant Name  \
0           0        3400025                             Jahanpanah   
1           1        3400341                    Rangrezz Restaurant   
2           2        3400005                Time2Eat - Mama Chicken   
3           3        3400021  Chokho Jeeman Marwari Jain Bhojanalya   
4           4        3400017                         Pinch Of Spice   

   Country Code  City  Longitude   Latitude  Average Cost for two  \
0             1  Agra  78.011544  27.161661                   850   
1             1  Agra   0.000000   0.000000                   700   
2             1  Agra  78.011608  27.160832                   500   
3             1  Agra  77.998092  27.195928                   400   
4             1  Agra  78.007553  27.201725                  1000   

   Has Table booking  Has Online delivery          ...           \
0                  0                    0          ...            
1                  0                    0          ...            
2                  0                    0          ...            
3                  0                    0          ...            
4                  0                    0          ...            

   Cuisine 1_Spanish  Cuisine 1_Steak  Cuisine 1_Street Food  Cuisine 1_Sushi  \
0                  0                0                      0                0   
1                  0                0                      0                0   
2                  0                0                      0                0   
3                  0                0                      0                0   
4                  0                0                      0                0   

   Cuisine 1_Tea  Cuisine 1_Tex-Mex  Cuisine 1_Thai  Cuisine 1_Tibetan  \
0              0                  0               0                  0   
1              0                  0               0                  0   
2              0                  0               0                  0   
3              0                  0               0                  0   
4              0                  0               0                  0   

   Cuisine 1_Turkish  Cuisine 1_Vietnamese  
0                  0                     0  
1                  0                     0  
2                  0                     0  
3                  0                     0  
4                  0                     0  

[5 rows x 90 columns]

In [53]:
cols = list(df.columns.values)
cols.pop(cols.index('Rating text'))
cols.pop(cols.index('Aggregate rating'))
dfClean = df[cols+['Rating text']+['Aggregate rating']] # Rearranging the Columns
dfClean.head()

Unnamed: 0  Restaurant ID                        Restaurant Name  \
0           0        3400025                             Jahanpanah   
1           1        3400341                    Rangrezz Restaurant   
2           2        3400005                Time2Eat - Mama Chicken   
3           3        3400021  Chokho Jeeman Marwari Jain Bhojanalya   
4           4        3400017                         Pinch Of Spice   

   Country Code  City  Longitude   Latitude  Average Cost for two  \
0             1  Agra  78.011544  27.161661                   850   
1             1  Agra   0.000000   0.000000                   700   
2             1  Agra  78.011608  27.160832                   500   
3             1  Agra  77.998092  27.195928                   400   
4             1  Agra  78.007553  27.201725                  1000   

   Has Table booking  Has Online delivery        ...         \
0                  0                    0        ...          
1                  0                    0        ...          
2                  0                    0        ...          
3                  0                    0        ...          
4                  0                    0        ...          

   Cuisine 1_Street Food  Cuisine 1_Sushi  Cuisine 1_Tea  Cuisine 1_Tex-Mex  \
0                      0                0              0                  0   
1                      0                0              0                  0   
2                      0                0              0                  0   
3                      0                0              0                  0   
4                      0                0              0                  0   

   Cuisine 1_Thai  Cuisine 1_Tibetan  Cuisine 1_Turkish  Cuisine 1_Vietnamese  \
0               0                  0                  0                     0   
1               0                  0                  0                     0   
2               0                  0                  0                     0   
3               0                  0                  0                     0   
4               0                  0                  0                     0   

   Rating text  Aggregate rating  
0            3               3.9  
1            3               3.5  
2            3               3.6  
3            4               4.0  
4            4               4.2  

[5 rows x 90 columns]

In [77]:
import numpy as np
x = dfClean.iloc[:,7:88] #Features
y = dfClean.iloc[:,88] #Target Value - Rating Text
X = np.array(x)
Y = np.array(y)
from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split( X, Y, test_size = 0.3, random_state = 1) #Splitting the data into test and train

## Building a Decision tree model because decision tree gave us the best accuracy previously

In [82]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score
dtree = DecisionTreeClassifier(min_samples_leaf=100)
model = dtree.fit(X_train, y_train)
dtree_pred = dtree.predict(X_test)
print('Decision Tree Accuracy:', accuracy_score(y_test, dtree_pred)*100)

Decision Tree Accuracy: 76.8104776579


### Changed the different criteria of the decision tree and increased the minimum samples in each leaf node of the tree
### Performing 10 fold cross validation to get the best accuracy

In [85]:
from sklearn.cross_validation import cross_val_score, cross_val_predict
from sklearn import metrics
scores = cross_val_score(model, X, Y, cv=10) #cv = 10 determines the k value for the k-fold cross validation
print ('Cross-validated scores:', scores*100)
print ('Avearge Accuracy:', np.mean(scores)*100)
print ('Best Accuracy:', max(scores)*100)

Cross-validated scores: [ 76.70126874  74.50980392  74.94226328  76.21247113  76.21247113
  76.90531178  75.26011561  74.88425926  76.94090382  75.63805104]
Avearge Accuracy: 75.820691972
Best Accuracy: 76.9409038239


## By changing the criteria and performing cross validation on the decision tree classifier, we obtained the best accuracy of almost 77%